### ISSUE TITLE
Emacs incorrectly warns about external modification when syncing in background. 

### ISSUE THREAD

**Issue by thblt** [cozy-labs/cozy-desktop#446]
When `cozy-desktop sync` is running and I work on a file in the Cozy directory, Emacs (24.5.1, latest packaged version, ArchLinux) keeps issuing warnings that the file has been externaly modified.  This seems to happen after each Emacs save (which happens often).

Something just a little bit weird seems to be happening with timestamps and Emacs doesn't like it at all.  I made the following test: 

 - Kill (nicely) `cozy-desktop` 
 - Modify a file in emacs.
 - `stat` the file:

    ```
 Accès : 2016-09-17 18:15:41.113809425 +0200
Modif. : 2016-09-17 18:15:41.113809425 +0200
Changt : 2016-09-17 18:15:41.120475957 +0200
  Créé : -
    ``` 
 - Run `cozy-desktop sync` again, wait until the file gets synced, kill it (nicely)
 - Re`stat`:

    ```
 Accès : 2016-09-17 18:16:41.315920663 +0200
Modif. : 2016-09-17 18:15:41.000000000 +0200
Changt : 2016-09-17 18:16:40.305941165 +0200
  Créé : -
    ```

**Comment by thblt** [comment]
(notice how *all* timestamps are different after the sync, when only `atime` should be)

**Comment by thblt** [comment]
Sorry for spamming, another precision.  I tried with an innocent file Emacs never touched.  I created it (`echo something > test`), `stat`, sync, stat, modified it, stat, sync,  stat.  The 4 `stat` outputs are:

```
# File just created
 Accès : 2016-09-17 18:29:51.973203695 +0200
Modif. : 2016-09-17 18:29:51.973203695 +0200
Changt : 2016-09-17 18:29:51.973203695 +0200
  Créé : -

# After initial sync
 Accès : 2016-09-17 18:30:15.419394405 +0200
Modif. : 2016-09-17 18:29:51.973203695 +0200
Changt : 2016-09-17 18:29:51.973203695 +0200
  Créé : -

# Modified
 Accès : 2016-09-17 18:30:15.419394405 +0200
Modif. : 2016-09-17 18:31:24.244663923 +0200
Changt : 2016-09-17 18:31:24.244663923 +0200
  Créé : -
 
# After second sync
 Accès : 2016-09-17 18:31:35.014445297 +0200
Modif. : 2016-09-17 18:31:24.000000000 +0200
Changt : 2016-09-17 18:31:34.004465799 +0200
  Créé : -
```

(Notice how every time, the fourth part of the `mtime` (nanoseconds? idk) gets set to 0)

**Comment by nono** [comment]
Hi,

thanks for the bug report.

> (Notice how every time, the fourth part of the mtime (nanoseconds? idk) gets set to 0)

Cozy-desktop keeps the time with only a precision of a second.

**Comment by thblt** [comment]
Until this is fixed on Cozy side, Emacs users may find the snippet below to be useful.  Simply yank [=paste] it somewhere in your `init.el` and Emacs will insure the buffer actually differs from its underlying file's contents *before* marking it as externally modified.  

It comes from [this StackOverflow answer](http://stackoverflow.com/a/29556894).  

~~~ elisp
;; Ignore modification-time-only changes in files, i.e. ones that
;; don't really change the contents.  This happens often with
;; switching between different VC buffers.

(defun update-buffer-modtime-if-byte-identical ()
  (let* ((size      (buffer-size))
         (byte-size (position-bytes size))
         (filename  buffer-file-name))
    (when (and byte-size (<= size 1000000))
      (let* ((attributes (file-attributes filename))
             (file-size  (nth 7 attributes)))
        (when (and file-size
                   (= file-size byte-size)
                   (string= (buffer-substring-no-properties 1 (1+ size))
                            (with-temp-buffer
                              (insert-file-contents filename)
                              (buffer-string))))
          (set-visited-file-modtime (nth 5 attributes))
          t)))))

(defun verify-visited-file-modtime--ignore-byte-identical (original &optional buffer)
  (or (funcall original buffer)
      (with-current-buffer buffer
        (update-buffer-modtime-if-byte-identical))))
(advice-add 'verify-visited-file-modtime :around #'verify-visited-file-modtime--ignore-byte-identical)

(defun ask-user-about-supersession-threat--ignore-byte-identical (original &rest arguments)
  (unless (update-buffer-modtime-if-byte-identical)
    (apply original arguments)))
(advice-add 'ask-user-about-supersession-threat :around #'ask-user-about-supersession-threat--ignore-byte-identical)
~~~

**Comment by sebn** [comment]
Hello there!
Sorry for the late answer.

Are you still getting this issue? With the same Emacs version and latest up-to-date Arch? And which cozy-desktop version are you using?

Thank you!

**Comment by thblt** [comment]
Hello sebn,

I've stopped using Cozy to manage most of my work files and reverted to a Git-based process, since Cozy desktop was getting really unstable with my increasingly large work folders (and even with regular versioned backups, it was getting a bit frightening).  (I haven't filed bugs because I really don't have time to try and reproduce seemingly random behavior, sorry about that.  Tell me if you're interested in more details :-)

Re the current bug, I see no commit referencing it.  If the probable cause of the problem (dropping nanoseconds when setting back the mtime) has been addressed, I can try to see if the problem with Emacs is solved.  If not, I can still try, but I don't believe in magic :-)

**Comment by sebn** [comment]
Well, I don't think the time precision issue was adressed, so you're probably right: no magic here :)

Thank you for giving it a try, anyway. The project should be more active during the upcoming months, with more platforms supported, and hopefully more bugs fixed. I'll keep this issue open and try to reproduce/fix it as soon as possible.

**Comment by thblt** [comment]
> Thank you for giving it a try, anyway.

OK, I'll do it when I have time.  Can't promise it will be soon though.

**Comment by sebn** [comment]
I meant "giving it a try in the past", I didn't expect you to test it forever when you don't use it anymore :)

Since this issue is part of the [known limitations](https://github.com/cozy-labs/cozy-desktop#limitations) of cozy-desktop, and we have tons of work to do, it's not really in the top list, so I can't promise either I'll fix it soon. I'll comment here when I can work on it.

**Comment by thblt** [comment]
> Since this issue is part of the known limitations of cozy-desktop

In fact, it is not listed in this section :-)  Adding it with a link to the [Emacs workaround](https://github.com/cozy-labs/cozy-desktop/issues/446#issuecomment-262239629) would be nice.

(I also noticed two typos in this section: "a reasonable size for files (< 1 To)" should be "1 Tb"; "git and other VCS are not meant to be share like this" should be something like "git (and other VCS) repositories are not meant to be shared this way")

**Comment by sebn** [comment]
Agreed. [Fixed](https://github.com/cozy-labs/cozy-desktop/commit/51b389f6935df0a37c4c661162185265319abef2) on master. Thanks!
